File: yanwte-spring/src/main/java/com/github/winteryoung/yanwte/spring/internals/ExtensionPointRegister.java
package com.github.winteryoung.yanwte.spring.internals;

import com.github.winteryoung.yanwte.YanwteException;
import com.github.winteryoung.yanwte.YanwteOptions;
import com.github.winteryoung.yanwte.spring.YanwteExtensionPoint;
import com.google.common.reflect.ClassPath;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.beans.factory.config.ConstructorArgumentValues;
import org.springframework.beans.factory.support.BeanDefinitionBuilder;
import org.springframework.beans.factory.support.DefaultListableBeanFactory;
import org.springframework.beans.factory.support.RootBeanDefinition;
import org.springframework.stereotype.Component;

/**
 * Auto register extension point providers after Spring initialization.
 *
 * @author Winter Young
 * @since 2016/10/22
 */
@SuppressWarnings("unused")
@Component
public class ExtensionPointRegister implements BeanFactoryPostProcessor {
    private Logger log = LoggerFactory.getLogger(getClass());

    @Override
    public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory)
            throws BeansException {
        DefaultListableBeanFactory defaultListableBeanFactory =
                (DefaultListableBeanFactory) beanFactory;
        List<Class<?>> classes;
        try {
            classes =
                    ClassPath.from(Thread.currentThread().getContextClassLoader())
                            .getAllClasses()
                            .stream()
                            .map(this::loadClass)
                            .filter(Objects::nonNull)
                            .collect(Collectors.toList());
        } catch (IOException e) {
            throw new YanwteException(e.getMessage(), e);
        }

        if (YanwteOptions.getLogExtensionsBuild() && log.isWarnEnabled()) {
            log.warn(
                    "Yanwte scanned classes:\n"
                            + classes.stream()
                                    .map(Class::getName)
                                    .collect(Collectors.joining("\n")));
        }

        List<Class<?>> extensionPointClasses =
                classes.stream()
                        .filter(
                                (Class<?> cls) -> {
                                    Annotation annotation = null;
                                    try {
                                        annotation = cls.getAnnotation(YanwteExtensionPoint.class);
                                    } catch (Throwable ignored) {
                                    }
                                    return annotation != null;
                                })
                        .collect(Collectors.toList());

        beanFactory.registerSingleton(
                "extensionPointProviderFactoryBean", new ExtensionPointProviderFactoryBean());

        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();

        for (Class<?> extensionPointClass : extensionPointClasses) {
            Class<?> providerClass = getProviderClass(extensionPointClass, classLoader);
            if (YanwteOptions.getLogExtensionsBuild() && log.isWarnEnabled()) {
                log.warn(
                        "Yanwte builds provider <"
                                + providerClass
                                + "> for interface <"
                                + extensionPointClass
                                + ">");
            }
            BeanDefinition beanDefinition = buildBeanDefinition(extensionPointClass, providerClass);
            String extensionPointClassName = extensionPointClass.getSimpleName();
            String uncapitalized =
                    extensionPointClassName.substring(0, 1).toLowerCase()
                            + extensionPointClassName.substring(1);

            defaultListableBeanFactory.registerBeanDefinition(
                    uncapitalized + "Provider", beanDefinition);
            defaultListableBeanFactory.clearMetadataCache();
        }
    }

    @Nullable
    private Class<?> loadClass(ClassPath.ClassInfo classInfo) {
        try {
            return classInfo.load();
        } catch (Throwable ex) {
            if (YanwteOptions.getLogExtensionsBuild() && log.isWarnEnabled()) {
                log.warn("Yanwte cannot load class: " + classInfo.getName(), ex);
            }
            return null;
        }
    }

    @NotNull
    private BeanDefinition buildBeanDefinition(
            Class<?> extensionPointClass, Class<?> providerClass) {
        RootBeanDefinition beanDefinition =
                (RootBeanDefinition)
                        BeanDefinitionBuilder.rootBeanDefinition(providerClass)
                                .setLazyInit(true)
                                .getRawBeanDefinition();
        beanDefinition.setFactoryBeanName("extensionPointProviderFactoryBean");
        beanDefinition.setFactoryMethodName("createExtensionPointProvider");
        ConstructorArgumentValues args = new ConstructorArgumentValues();
        args.addGenericArgumentValue(extensionPointClass);
        beanDefinition.setConstructorArgumentValues(args);
        beanDefinition.setTargetType(extensionPointClass);
        return beanDefinition;
    }

    private Class<?> getProviderClass(Class<?> extensionPointClass, ClassLoader classLoader) {
        try {
            return classLoader.loadClass(extensionPointClass.getName() + "Provider");
        } catch (ClassNotFoundException e) {
            throw new YanwteException(e.getMessage(), e);
        }
    }
}

File: yanwte-spring/src/main/java/com/github/winteryoung/yanwte/spring/internals/ExtensionPointRegister.kt
package com.github.winteryoung.yanwte.spring.internals

import com.github.winteryoung.yanwte.YanwteException
import com.github.winteryoung.yanwte.YanwteOptions
import com.github.winteryoung.yanwte.internals.utils.ReflectionUtils
import com.github.winteryoung.yanwte.spring.YanwteExtensionPoint
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.BeansException
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.beans.factory.config.BeanFactoryPostProcessor
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory
import org.springframework.beans.factory.config.ConstructorArgumentValues
import org.springframework.beans.factory.support.BeanDefinitionBuilder
import org.springframework.beans.factory.support.DefaultListableBeanFactory
import org.springframework.beans.factory.support.RootBeanDefinition
import org.springframework.stereotype.Component
import java.io.IOException

/**
 * Auto register extension point providers after Spring initialization.

 * @author Winter Young
 * *
 * @since 2016/10/22
 */
@Component
class ExtensionPointRegister : BeanFactoryPostProcessor {
    private val log = LoggerFactory.getLogger(javaClass)

    @Throws(BeansException::class)
    override fun postProcessBeanFactory(beanFactory: ConfigurableListableBeanFactory) {
        val defaultListableBeanFactory = beanFactory as DefaultListableBeanFactory
        val classes: Array<Class<*>>
        val classLoader = Thread.currentThread().contextClassLoader

        try {
            classes = ReflectionUtils.getClasses("", classLoader)
        } catch (e: IOException) {
            throw YanwteException(e.message, e)
        }

        val extensionPointClasses = classes.filter { cls: Class<*> ->
                cls.getAnnotation(YanwteExtensionPoint::class.java) != null
        }

        beanFactory.registerSingleton("extensionPointProviderFactoryBean", ExtensionPointProviderFactoryBean())

        for (extensionPointClass in extensionPointClasses) {
            val providerClass = getProviderClass(extensionPointClass, classLoader)
            if (YanwteOptions.logExtensionsBuild && log.isWarnEnabled) {
                log.warn("postProcessBeanFactory, extensionPointInterface: " + extensionPointClass
                        + ", provider: " + providerClass)
            }
            val beanDefinition = buildBeanDefinition(extensionPointClass, providerClass)
            val uncapitalized = StringUtils.uncapitalize(extensionPointClass.simpleName)

            defaultListableBeanFactory.registerBeanDefinition(uncapitalized + "Provider", beanDefinition)
            defaultListableBeanFactory.clearMetadataCache()
        }
    }

    private fun buildBeanDefinition(extensionPointClass: Class<*>, providerClass: Class<*>): BeanDefinition {
        val beanDefinition = BeanDefinitionBuilder
                .rootBeanDefinition(providerClass)
                .setLazyInit(true)
                .rawBeanDefinition as RootBeanDefinition
        beanDefinition.factoryBeanName = "extensionPointProviderFactoryBean"
        beanDefinition.factoryMethodName = "createExtensionPointProvider"
        val args = ConstructorArgumentValues()
        args.addGenericArgumentValue(extensionPointClass)
        beanDefinition.constructorArgumentValues = args
        beanDefinition.targetType = extensionPointClass
        return beanDefinition
    }

    private fun getProviderClass(extensionPointClass: Class<*>, classLoader: ClassLoader): Class<*> {
        try {
            return classLoader.loadClass(extensionPointClass.name + "Provider")
        } catch (e: ClassNotFoundException) {
            throw YanwteException(e.message, e)
        }
    }
}
